src/middleware/validation.middleware.js

```diff
@@ -0,0 +1,56 @@
+import Joi from 'joi';
+import Schemas from './validation.schema';
+
+export default () => {
+  // Joi validation options
+  const validationOptions = {
+    abortEarly: false,
+    allowUnknown: true,
+    stripUnknown: true
+  };
+
+  // return the validation middleware
+  return (req, res, next) => {
+    const route = req.route.path;
+
+    if (route in Schemas) {
+      // get schema for the current route
+      const schema = Schemas[route];
+      const toValidate = {};
+
+      if (!schema) {
+        return next();
+      }
+
+      ['params', 'body', 'query'].forEach(key => {
+        if (schema[key]) {
+          toValidate[key] = req[key];
+        }
+      });
+
+      return Joi.validate(
+        toValidate,
+        schema,
+        validationOptions,
+        (err, data) => {
+          if (err) {
+            // Custom Error
+            const SimplifiedError = {
+              status: 'validation error',
+              message: err.details
+                ? err.details[0].message.replace(/['"]/g, '')
+                : err.message
+            };
+            // Send back the JSON error response
+            res.status(400).json(SimplifiedError);
+          } else {
+            // Replace req.body with the data after Joi validation
+            req = data;
```

Comment:
`req` is not same as `req.body`. You're overwriting the whole HTTP request

Reply:
Sorry, the comment was meant to `Replace req with the data after Joi validation`. The idea was to overwrite the request since Joi might be changing some data after validation. 
Maybe a better approach would have been 
`req.params = data.params
 req.body = data.body
 req.query = data.query
 req.headers = data.headers
`

Reply:
The request object contains other methods and interfaces that are required by the framework to do certain things, so just update the part of the request that you need and leave the rest intact

Reply:
Okay. Thank you